### PR TITLE
Make fetch shuffle partition data in parallel

### LINKED ISSUE #208: Make executor fetch shuffle partition data in parallel
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

Currently, executors fetch shuffle partition data from upstreams sequentially, which is not efficient.

**Describe the solution you'd like**

It's better to make it run in parallel with some throughput control. What's more, we can optimize the shuffle data fetching way to avoid hot executor IOs.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by yahoNanJing**
# Which issue does this PR close?

Closes #208.

 # Rationale for this change

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by yahoNanJing** [comment]
Hi @andygrove, @thinkharderdev, @avantgardnerio, could you help review this PR?

File: ballista/rust/core/src/execution_plans/shuffle_reader.rs [hunk 1]
```diff
@@ -102,29 +107,35 @@ impl ExecutionPlan for ShuffleReaderExec {
...
+        }
+        // Sort partitions for evenly send fetching partition requests to avoid hot executors within one task
+        let mut partition_locations: Vec<PartitionLocation> = partition_locations
+            .into_values()
+            .flat_map(|ps| ps.into_iter().enumerate())
+            .sorted_by(|(p1_idx, _), (p2_idx, _)| Ord::cmp(p1_idx, p2_idx))
+            .map(|(_, p)| p)
+            .collect();
+        // Shuffle partitions for evenly send fetching partition requests to avoid hot executors within multiple tasks
+        partition_locations.shuffle(&mut thread_rng());
```
Comment (thinkharderdev) [inline ballista/rust/core/src/execution_plans/shuffle_reader.rs:132]: Why do we sort by index before shuffling here?

**Review by thinkharderdev — APPROVED** [review]
One question but otherwise LGTM

File: ballista/rust/core/src/execution_plans/shuffle_reader.rs [hunk 2]
```diff
@@ -102,29 +107,35 @@ impl ExecutionPlan for ShuffleReaderExec {
...
-
-                r.map_err(|e| ArrowError::ExternalError(Box::new(e)))
-            })
-            .try_flatten()
-        });
+        let task_id = context.task_id().unwrap_or_else(|| partition.to_string());
+        info!("ShuffleReaderExec::execute({})", task_id);
+
+        // TODO make the maximum size configurable, or make it depends on global memory control
+        let max_request_num = 50usize;
```
Comment (andygrove) [inline ballista/rust/core/src/execution_plans/shuffle_reader.rs:116]: +1 for making this a config option that is documented in the user guide. I can help with this in a follow-on PR.

File: ballista/rust/core/src/execution_plans/shuffle_reader.rs (hunk 1 above)
Comment (yahoNanJing) [reply ballista/rust/core/src/execution_plans/shuffle_reader.rs:132]: Sorting first may be helpful for reducing the bias of random chosen. Maybe it's not necessary.

File: ballista/rust/core/src/execution_plans/shuffle_reader.rs [hunk 3]
```diff
@@ -102,29 +107,35 @@ impl ExecutionPlan for ShuffleReaderExec {
...
-                let timer = fetch_time.timer();
-                let r = fetch_partition(&p).await;
-                timer.done();
-
-                r.map_err(|e| ArrowError::ExternalError(Box::new(e)))
-            })
-            .try_flatten()
-        });
+        let task_id = context.task_id().unwrap_or_else(|| partition.to_string());
+        info!("ShuffleReaderExec::execute({})", task_id);
```
Comment (andygrove) [inline ballista/rust/core/src/execution_plans/shuffle_reader.rs:113]: This could make the logs quite noisy again?

**Review by andygrove — APPROVED** [review]
LGTM. Thanks @yahoNanJing!

**Comment by mingmwang** [comment]
@yahoNanJing Could you please explain a little about the error handling case ?
For example, if one of the parallel fetch failed, what happened to the others ?
And what happened if there are multiple parallel fetch failures, will all the error info correctly propagated to the stream consumer ?

**Comment by yahoNanJing** [comment]
@mingmwang Just added one commit to abort fast when error occurs.

**Pull request merged**